File: solutions/c/matching-brackets/1/matching_brackets.c
#include "matching_brackets.h"

struct Stack {
    int data[100];
    int top;
};

static int push(struct Stack *stack, char d) {
    stack->data[stack->top++] = d;
    return 1;
}

static int pop(struct Stack *stack) {
    if (stack->top == 0) {
        return -1;
    }
    int d = stack->data[--stack->top];

    return d;
}

bool is_paired(const char *input) {
    struct Stack parenStack = {{0}, 0};

    for (; *input; input++) {
        if (*input == '{' || *input == '(' || *input == '[') {
            push(&parenStack, *input);
        } else if (*input == '}' || *input == ')' || *input == ']') {
            int success = pop(&parenStack);
            if (success == -1) {
                return false;
            } else if (success + 1 != *input && success + 2 != *input) {
                return false;
            }
        }
    }

    return parenStack.top == 0;
}

File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string: str) -> bool:
    stack = []

    for c in input_string:
        if c in "{[(":
            stack.append(c)
        elif c in "}])":
            if len(stack) == 0:
                return False
            popped = stack.pop()
            if ord(popped) + 2 != ord(c) and ord(popped) + 1 != ord(c):
                return False

    return len(stack) == 0
